src/wallet/rpcwallet.cpp

```diff
@@ -597,6 +602,92 @@ UniValue listaddresses(const UniValue& params, bool fHelp)
         entry.pushKV("source", "legacy_hdseed");
 
         bool hasData = add_sapling(saplingAddresses, PaymentAddressSource::LegacyHDSeed, entry);
+        if (hasData) {
+            ret.push_back(entry);
+        }
+    }
+
+    // mnemonic seed source
+    {
+        UniValue entry(UniValue::VOBJ);
+        entry.pushKV("source", "mnemonic_seed");
+
+        bool hasData = false;
+        // transparent
+        // For each address in pwalletMain->mapKeyMetadata, include any address
+        // for which its HD keypath metadata indicates it is associated with the legacy
+        // account ID. All such addresses are derived from the mnemonic seed.
+        UniValue mnemonic_taddrs(UniValue::VARR);
+        for (const auto& [keyId, keyMeta] : pwalletMain->mapKeyMetadata) {
+            auto account = libzcash::ParseHDKeypathAccount(32, Params().BIP44CoinType(), keyMeta.hdKeypath);
+            if (account.has_value() && account.value() == ZCASH_LEGACY_ACCOUNT) {
+                mnemonic_taddrs.push_back(keyIO.EncodeDestination(keyId));
+                hasData = true;
+            }
+        }
+        if (hasData) {
+            entry.pushKV("transparent", mnemonic_taddrs);
+        }
+
+        // sapling
+        hasData |= add_sapling(saplingAddresses, PaymentAddressSource::MnemonicHDSeed, entry);
+
+        // unified
+        // here, we want to use the information in mapUfvkAddressMetadata to report all the unified addresses
+        UniValue unified_groups(UniValue::VARR);
+        auto hdChain = pwalletMain->GetMnemonicHDChain();
+        for (const auto& [ufvkid, addrmeta] : pwalletMain->mapUfvkAddressMetadata) {
+            auto account = pwalletMain->GetUnifiedAccountId(ufvkid);
+            if (account.has_value() && hdChain.has_value()) {
+                // double-check that the ufvkid we get from address metadata is actually
+                // associated with the mnemonic HD chain
+                auto ufvkCheck = pwalletMain->mapUnifiedAccountKeys.find(
+                    std::make_pair(hdChain.value().GetSeedFingerprint(), account.value())
+                );
+                if (ufvkCheck != pwalletMain->mapUnifiedAccountKeys.end() && ufvkCheck->second == ufvkid) {
+                    auto hdKeyPath = Zip32AccountKeyPath(Params().BIP44CoinType(), account.value());
+                    UniValue unified_group(UniValue::VOBJ);
+                    unified_group.pushKV("zip32KeyPath", hdKeyPath);
```

Comment:
Hmm, I don't think we want `zip32KeyPath` for Unified Addresses, because it only applies to the Sapling (and I guess Orchard) keys within the UA (transparent will use the BIP 44 keypath). I think we should instead have an `account` field containing the account number. We should maybe also have a `seedfp` field with the hex of the mnemonic seed (though that's really only necessary if we had multiple mnemonics, so we could leave this field out for now I guess).

Reply:
If we think `seedfp` will be necessary then we should add it now, to simplify things for future consumers that would need to handle the case where it's missing.